### PR TITLE
Add FlaxWhisperForAudioClassification model

### LINKED ISSUE #21779: Add Flax Whisper for audio classification
### Feature request

The PR https://github.com/huggingface/transformers/pull/21754 adds the PyTorch version of WhisperForAudioClassification. It would be great to add the Flax equivalent for cross-library equivalence ♻️

### Motivation

Whisper is an encoder-decoder model for speech recognition. However, we can repurpose the model for other speech tasks, such as audio classification.

Audio classification is the task of mapping from an input speech sequence to a single class prediction. For more details, refer to the task page on the Hub: https://huggingface.co/tasks/audio-classification

For audio classification, we only require a single model output. Thus, we do not need the auto-regressive generation capacities of the Whisper decoder (which is used to generate a sequence of text tokens during speech recognition). Instead, we can just use the Whisper encoder to get hidden states, and add a classification head on top to make class label predictions.

This is analogous to using a Wav2Vec2 model for audio classification: the Wav2Vec2 encoder is used to get hidden states, and a classification head added on top to make class label predictions.

The PR https://github.com/huggingface/transformers/pull/21754 adds the PyTorch version of WhisperForAudioClassification. It required adding a projection layer and classification layer on top of the WhisperEncoder. For more details, refer directly to the pull request.

It would be great to add the Flax equivalent of this model for cross-framework support.

The most difficult part of this PR will be getting the model tester to work. You can see from the PyTorch PR that we require a standalone tester for the audio classification model. This is because the original Whisper model is an encoder-decoder model, but the audio classification model is an encoder-only model. Thus, we require different testing logic.

### Your contribution

Opening this one up to the community! This will be quite a fun JAX/Flax PR! 🚀

If you're interested in tackling this, free to drop a comment in this thread and open a PR when you're ready. More than happy to answer any questions / queries about this integration!

### PR DESCRIPTION
**Pull request by raghavanone**
# What does this PR do?

Fix : #21779 

Please review and let me know changes @sanchit-gandhi

### CODE REVIEW

**Review by sanchit-gandhi — COMMENTED** [review]
Modelling code looks good @raghavanone! Nice one on getting this working so quickly 🙌 Do you want to have a go at adding the encoder-only tests? See the PyTorch WhisperForAudioClassficiation PR for details, think you can also add these quite quickly :)

**Comment by raghavanone** [comment]
> Modelling code looks good @raghavanone! Nice one on getting this working so quickly 🙌 Do you want to have a go at adding the encoder-only tests? See the PyTorch WhisperForAudioClassficiation PR for details, think you can also add these quite quickly :)

I have added the Encoder tests, But some test are failing, The FlaxWhisperForAudioClassification class extends FlaxWhisperPreTrainedModel . Due to this inheritance, the call method expects decoder related params. 

Should the FlaxWhisperForAudioClassification not extend FlaxWhisperPreTrainedModel instead create a new pretrainedclass ?

**Comment by sanchit-gandhi** [comment]
Hey @raghavanone! The PyTorch model has just been merged (https://github.com/huggingface/transformers/pull/21754), so you can rebase onto main to get the required config changes:
```
git fetch upstream
git rebase upstream/main
```
This will fix the failing Flax tests we're getting here: https://app.circleci.com/pipelines/github/huggingface/transformers/58972/workflows/2388bd70-553e-412f-9ee7-0599cace5639/jobs/719829

The only thing to make sure is that the first time you push after rebasing, you **force push** to origin:
```
git add .
git commit -m "Some new changes after rebase"
git push -f origin fix_issue_21779
```

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 1]
```diff
@@ -1468,3 +1469,123 @@ def update_inputs_for_generation(self, model_outputs, model_kwargs):
...
+        self.projector = nn.Dense(self.config.classifier_proj_size)
+        self.classifier = nn.Dense(self.config.num_labels, dtype=self.dtype)
+
+    def __call__(
+        self,
+        input_features,
+        encoder_outputs=None,
+        output_attenions=None,
+        output_hidden_states: bool = True,
+        decoder_input_ids=None,
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:1493]: We don't need `decoder_input_ids` for an encoder only model! 😉

**Comment by raghavanone** [comment]
@sanchit-gandhi  There are 2 test failing here, I am unable to get the same failure locally in my machine. Any pointers on how to replicate failing test and fix it ?

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 2]
```diff
@@ -1468,3 +1469,177 @@ def update_inputs_for_generation(self, model_outputs, model_kwargs):
...
+
+    Transcription example:
+
+    ```python
+    >>> from transformers import WhisperProcessor, FlaxWhisperForAudioClassification
+    >>> from datasets import load_dataset
+
+    >>> processor = WhisperProcessor.from_pretrained("openai/whisper-tiny.en")
+    >>> model = FlaxWhisperForAudioClassification.from_pretrained("openai/whisper-tiny.en", from_pt=True)
+    >>> ds = load_dataset("hf-internal-testing/librispeech_asr_dummy", "clean", split="validation")
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:1632]: We need to update this to an audio classification example! Feel free to copy the one from the PyTorch code and adapt:

```python
import jax.numpy as jnp
from transformers import AutoFeatureExtractor, FlaxWhisperForAudioClassification
from datasets import load_dataset

feature_extractor = AutoFeatureExtractor.from_pretrained("sanchit-gandhi/whisper-medium-fleurs-lang-id")
model = FlaxWhisperForAudioClassification.from_pretrained("sanchit-gandhi/whisper-medium-fleurs-lang-id", from_pt=True)
ds = load_dataset("google/fleurs", "all", split="validation", streaming=True)

sample = next(iter(ds))

inputs = feature_extractor(
    sample["audio"]["array"], sampling_rate=sample["audio"]["sampling_rate"], return_tensors="np"
)
input_features = inputs.input_features

logits = model(input_features).logits

predicted_class_ids = jnp.argmax(logits).item()
predicted_label = model.config.id2label[predicted_class_ids]
predicted_label
```
```
'af_za'
```

File: src/transformers/models/whisper/modeling_flax_whisper.py [hunk 3]
```diff
@@ -1468,3 +1469,177 @@ def update_inputs_for_generation(self, model_outputs, model_kwargs):
...
+            return encode_module(input_features, **kwargs)
+
+        return self.module.apply(
+            {"params": params or self.params},
+            input_features=jnp.array(input_features, dtype="f4"),
+            output_attentions=output_attentions,
+            output_hidden_states=output_hidden_states,
+            return_dict=return_dict,
+            rngs=rngs,
+            # method=_encoder_forward,
```
Comment (sanchit-gandhi) [inline src/transformers/models/whisper/modeling_flax_whisper.py:1617]: Can remove this commented line too

File: tests/models/whisper/test_modeling_flax_whisper.py [hunk 4]
```diff
@@ -704,3 +703,232 @@ def test_tiny_timestamp_generation(self):
 
         transcript = processor.batch_decode(generated_ids, skip_special_tokens=True, output_offsets=True)
         self.assertEqual(transcript, EXPECTED_TRANSCRIPT)
+
+
+def prepare_whisper_encoder_inputs_dict(
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_flax_whisper.py:708]: This function is unused - can remove

File: tests/models/whisper/test_modeling_flax_whisper.py [hunk 5]
```diff
@@ -704,3 +703,232 @@ def test_tiny_timestamp_generation(self):
...
+class WhisperEncoderModelTest(FlaxModelTesterMixin, unittest.TestCase):
+    all_model_classes = (FlaxWhisperForAudioClassification,) if is_flax_available() else ()
+    is_encoder_decoder = False
+    fx_compatible = False
+    test_pruning = False
+    test_missing_keys = False
+
+    input_name = "input_features"
+
+    # def setUp(self):
```
Comment (sanchit-gandhi) [inline tests/models/whisper/test_modeling_flax_whisper.py:847]: More commented lines we can remove!

**Review by sanchit-gandhi — COMMENTED** [review]
Nice work @raghavanone! Mainly just some clean-up before we can get this merged!

**Review by sanchit-gandhi — COMMENTED** [review]
Nice one @raghavanone! Mainly just some code clean-up, then we can get this merged!

**Comment by sanchit-gandhi** [comment]
Hey @raghavanone! Would you mind going through the previous review comments and marking them as resolved where you've addressed them? I'll then get you a final review asap! Thanks!

**Comment by sanchit-gandhi** [comment]
Hey @raghavanone - I think the commit history has been corrupted for this PR? Gentle reminder that one must force push after rebasing: https://github.com/huggingface/transformers/pull/21894#issuecomment-1458359220 Think this is probably the culprit for the 250 extra commits!

In this instance, it's probably best to close this PR in favour of a new one that only contains the new changes you with to merge. Sorry about that!

**Comment by sanchit-gandhi** [comment]
Closing in favour of #22883
